meta/bucket.go

```diff
@@ -61,10 +78,102 @@ func (m *Meta) GetBucketInfo(bucketName string) (Bucket, error) {
 }
 
 func (m *Meta) GetUserInfo(uid string) ([]string, error) {
-	m.Cache.Remove(redis.UserTable, uid)
+	m.Cache.Remove(redis.UserTable, USER_CACHE_PREFIX, uid)
 	buckets, err := m.GetUserBuckets(uid, true)
 	if err != nil {
 		return nil, err
 	}
 	return buckets, nil
 }
+
+/*
+* init bucket usage cache when meta is newed.
+*
+ */
+func (m *Meta) InitBucketUsageCache() error {
+	// the map contains the bucket usage which are not in cache.
+	bucketUsageMap := make(map[string]*Bucket)
+	// the map contains the bucket usage which are in cache and will be synced into database.
+	bucketUsageCacheMap := make(map[string]int64)
+	// the usage in buckets table is accurate now.
+	buckets, err := m.Client.GetBuckets()
+	if err != nil {
+		helper.Logger.Println(2, "failed to get buckets from db. err: ", err)
+		return err
+	}
+
+	// init the bucket usage key in cache.
+	for _, bucket := range buckets {
+		bucketUsageMap[bucket.Name] = bucket
+	}
+
+	// try to get all bucket usage keys from cache.
+	pattern := fmt.Sprintf("%s*", BUCKET_CACHE_PREFIX)
+	bucketsInCache, err := m.Cache.Keys(redis.BucketTable, pattern)
+	if err != nil {
+		helper.Logger.Println(2, "failed to get bucket usage from cache, err: ", err)
+		return err
+	}
+
+	if len(bucketsInCache) > 0 {
+		// query all usages from cache.
+		for _, bic := range bucketsInCache {
+			usage, err := m.Cache.HGetInt64(redis.BucketTable, BUCKET_CACHE_PREFIX, bic, FIELD_NAME_USAGE)
+			if err != nil {
+				helper.Logger.Println(2, "failed to get usage for bucket: ", bic, " with err: ", err)
+				continue
+			}
+			// add the to be synced usage.
+			bucketUsageCacheMap[bic] = usage
+			if _, ok := bucketUsageMap[bic]; ok {
+				// if the key already exists in cache, then delete it from map
+				delete(bucketUsageMap, bic)
+			}
+		}
+
+	}
+
+	// init the bucket usage in cache.
+	if len(bucketUsageMap) > 0 {
+		for _, bk := range bucketUsageMap {
+			fields, err := bk.Serialize()
+			if err != nil {
+				helper.Logger.Println(2, "failed to serialize for bucket: ", bk.Name, " with err: ", err)
+				return err
+			}
+			_, err = m.Cache.HMSet(redis.BucketTable, BUCKET_CACHE_PREFIX, bk.Name, fields)
+			if err != nil {
+				helper.Logger.Println(2, "failed to set bucket to cache: ", bk.Name, " with err: ", err)
+				return err
+			}
+		}
+
+	}
+	// sync the buckets usage in cache into database.
+	if len(bucketUsageCacheMap) > 0 {
+		err = m.Client.UpdateUsages(bucketUsageCacheMap, nil)
+		if err != nil {
+			helper.Logger.Println(2, "failed to sync usages to database, err: ", err)
+			return err
+		}
+	}
+	return nil
+}
+
+func (m *Meta) bucketUsageSync(bucketName string) error {
```

Comment:
Maybe add `usage` value to sync event message so no need to check from Redis again, meanwhile other improvements, like merging consecutive writes, become possible.

Reply:
thanks @cannium for your good suggestions, i will encapsulate the bucket name and it's usage into the parameter and update it to db directly.
Consecutive writes is a good performance improvements suggestion. Can we make enhancement for it in the next step? :)

Reply:
Sure, do as your convenience.

Reply:
@cannium, please help to review this PR again. I have amended the bucket sync logic and just transfer the bucket name and usage to update db directly. Thanks.